org.scala-ide.sdt.core.tests/src/org/scalaide/core/sbtbuilder/NonScalaSourceErrorMarkersTest.scala

```diff
@@ -0,0 +1,81 @@
+package org.scalaide.core
+package sbtbuilder
+
+import org.eclipse.core.resources.IProject
+import org.eclipse.core.resources.IProjectDescription
+import org.eclipse.core.resources.IncrementalProjectBuilder
+import org.junit.Assert
+import org.junit.Ignore
+import org.junit.Test
+import org.scalaide.core.SdtConstants
+import org.scalaide.ui.internal.preferences.ScalaPluginSettings
+import org.scalaide.util.internal.SettingConverterUtil
+
+import testsetup.Implicits
+import testsetup.SDTTestUtils
+
+class NonScalaSourceErrorMarkersTest {
+  import SDTTestUtils._
+  import Implicits.TestableProject
+
+  def addScalaNature(project: IProject) = natures(project) { description =>
+    SdtConstants.NatureId +: description.getNatureIds
+  }
+
+  private def natures(project: IProject)(newNatures: IProjectDescription => Array[String]): Unit = {
+    val description = project.getDescription
+    description.setNatureIds(newNatures(description))
+    project.setDescription(description, null)
+  }
+
+  def removeScalaNature(project: IProject) = natures(project) { description =>
+    description.getNatureIds.filter { _ != SdtConstants.NatureId }
+  }
+
+  def soThen(assertThat: => Unit): Unit = assertThat
+
+  @Test def shouldFindJavaErrorMarkerTooWhenScalaNatureOfAisOffAndOn(): Unit = {
+    val Seq(prjA, prjB, prjC) = createProjects("A", "B", "C")
+
+    try {
+      prjB dependsOnAndExports prjA
+      prjC onlyDependsOn prjB
+
+      val Seq(packA, packB, packC) = Seq(prjA, prjB, prjC).map(createSourcePackage("test"))
+
+      val unitA = packA.createCompilationUnit("A.scala", "trait A", true, null)
+      val unitB = packB.createCompilationUnit("B.scala", "class B extends A", true, null)
+      val unitC = packC.createCompilationUnit("C.scala", "class C extends B", true, null)
+
+      // set stopOnBuild to true
+      val stopBuildOnErrors = SettingConverterUtil.convertNameToProperty(ScalaPluginSettings.stopBuildOnErrors.name)
+      IScalaPlugin().getPreferenceStore.setValue(stopBuildOnErrors, true)
+
+      // no errors
+      SDTTestUtils.workspace.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, null)
+      val unitsToWatch = Seq(unitA, unitB, unitC)
+      val errors = SDTTestUtils.getErrorMessages(unitsToWatch: _*)
+      Assert.assertEquals("No build errors", Seq(), errors)
+
+      def whenRemoveAndAddScalaNatureToA(): Unit = {
+        removeScalaNature(prjA.underlying)
+        SDTTestUtils.workspace.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, null)
+        addScalaNature(prjA.underlying)
+        SDTTestUtils.workspace.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, null)
+      }
+      whenRemoveAndAddScalaNatureToA()
```

Comment:
Ok, now I understand why you introduced this `soThen` method. Kind of tricky to understand such a multi method description of a test case.